src/pinghelper.cpp

```diff
@@ -31,53 +50,141 @@ void PingHelper::start(const QString& serverIpv4Gateway,
 
   m_gateway = serverIpv4Gateway;
   m_source = deviceIpv4Address.section('/', 0, 0);
+  m_pingSender =
+#if defined(MVPN_LINUX) || defined(MVPN_ANDROID)
+      new LinuxPingSender(m_source, this);
+#elif defined(MVPN_MACOS) || defined(MVPN_IOS)
+      new MacOSPingSender(m_source, this);
+#elif defined(MVPN_WINDOWS)
+      new WindowsPingSender(m_source, this);
+#else
+      new DummyPingSender(m_source, this);
+#endif
+  connect(m_pingSender, &PingSender::recvPing, this, &PingHelper::pingReceived);
+
+  // Reset the ping statistics
+  m_sequence = 0;
+  for (int i = 0; i < PING_STATS_WINDOW; i++) {
+    m_pingData[i].timestamp = -1;
+    m_pingData[i].latency = -1;
+    m_pingData[i].sequence = 0;
+  }
+
   m_pingTimer.start(PING_TIMOUT_SEC * 1000);
 }
 
 void PingHelper::stop() {
   logger.log() << "PingHelper deactivated";
 
+  if (m_pingSender) {
+    delete m_pingSender;
+    m_pingSender = nullptr;
+  }
+
   m_pingTimer.stop();
+}
 
-  for (PingSender* pingSender : m_pings) {
-    pingSender->deleteLater();
+void PingHelper::nextPing() {
+#ifdef QT_DEBUG
+  logger.log() << "Sending ping seq:" << m_sequence;
+#endif
+
+  int index = m_sequence % PING_STATS_WINDOW;
+  m_pingData[index].timestamp = QDateTime::currentMSecsSinceEpoch();
+  m_pingData[index].latency = -1;
+  m_pingData[index].sequence = m_sequence;
+  m_pingSender->sendPing(m_gateway, m_sequence);
+
+  m_sequence++;
```

Comment:
1 line of comment saying that we do not care about the overflow of this variable because we use it as an index + a counter.